ieee1905_transport: use bind() instead of bpf filter

fvbogaert opened the pull request:
The IEEE1905 transport opens a raw socket to capture 1905.1 and LLDP packets. It currently uses a BPF filter to do that. However, on some platforms, a BPF filter causes the acceleration fabric to stop working or to degrade.

A better solution is to use bind() to tie the raw socket to a specific ethernet protocol. This looks more or less like this:
```c
int openPacketSocket(int ifindex, uint16_t eth_type)
{
    int                 s;
    struct sockaddr_ll  socket_address;

    s = socket(AF_PACKET, SOCK_RAW, htons(eth_type));
    if (-1 == s)
    {
        return -1;
    }

    memset(&socket_address, 0, sizeof(socket_address));
    socket_address.sll_family   = AF_PACKET;
    socket_address.sll_ifindex  = ifindex;
    socket_address.sll_protocol = htons(eth_type);

    if (-1 == bind(s, (struct sockaddr*)&socket_address, sizeof(socket_address)))
    {
        close(s);
        return -1;
    }

    return s;
}
```

Note that a socket can be bound only to 1 ethertype. Therefore, we need to use 2 sockets per interface instead of just one.

fvb-mind committed:
transport: ieee1905_transport: replace BPF filter with bind()

The IEEE1905 transport opens a raw socket to capture 1905.1 and LLDP packets.
It currently uses a BPF filter to do that.
However, on some platforms, a BPF filter causes the acceleration fabric to
stop working or to degrade.

Instead, bind() is used to tie the socket to specific protocols.
Note that two sockets are now needed instead of one (one for each
protocol).

Signed-off-by: Frederik Van Bogaert <frederik.vanbogaert@essensium.com>

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Replace BPF filter in transport](https://trello.com/c/MMvXUWzL/55-replace-bpf-filter-in-transport)

fvb-mind committed:
transport: ieee1905-transport: revert to SOCK_RAW i.s.o. SOCK_DGRAM

This is a task for another time, in the mean time it was causing the
test to fail.

Signed-off-by: Frederik Van Bogaert <frederik.vanbogaert@essensium.com>

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
-    if ((sockfd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ALL))) < 0) {
-        MAPF_ERR("cannot open raw socket \"" << strerror(errno) << "\" (" << errno << ").");
-        return false;
+    if ((sockfd = socket(AF_PACKET, SOCK_DGRAM, htons(ether_type))) < 0) {
```
We still want SOCK_RAW, otherwise the Ethernet header is stripped off and we parse that.

Note that it's not strictly needed (we can get source and destination address through ancillary data) but it's a lot simpler that way.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
-        MAPF_ERR("cannot set socket option SO_REUSEADDR \"" << strerror(errno) << "\" (" << errno
-                                                            << ").");
+        MAPF_ERR("cannot set socket option SO_REUSEADDR (errno: " << errno << " ["
+                                                                  << strerror(errno) << "])");
```
This is an unrelated change. Please do such changes in a separate commit. (nitpick)

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
+{
+    MAPF_DBG("opening raw socket on interface " << if_index << ".");
+
+    if (network_interfaces_[if_index].ieee1905_socket != -1) {
```
This should be in two separate conditions.

though actually, the condition is redundant: close(-1) will return an error, but we ignore its errors anyway.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
+    int lldp_socket     = open_packet_socket(if_index, ETHERTYPE_LLDP);
 
+    if (ieee1905_socket < 0 || lldp_socket < 0) {
+        // if only the second call failed, the first socket still needs to be close()d
```
And vice-versa. Again, since close(-1) is OK, just close both.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -137,50 +81,62 @@ void Ieee1905Transport::update_network_interfaces(
     }
 
-    network_interfaces_[if_index].fd = sockfd;
+    int ieee1905_socket = open_packet_socket(if_index, ETHERTYPE_IEEE1905);
+    int lldp_socket     = open_packet_socket(if_index, ETHERTYPE_LLDP);
```
There is no need for temporary variables, we can write directly to network_interfaces_[if_index].ieee1905_socket.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
-        network_interfaces_[if_index].fd = -1;
+    if (!is_active && network_interfaces_[if_index].ieee1905_socket >= 0) {
+        poller_.Remove(network_interfaces_[if_index].ieee1905_socket);
+        poller_.Remove(network_interfaces_[if_index].lldp_socket);
```
Condition should be duplicated for lldp_socket

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
-            poller_.Add(network_interfaces_[if_index].fd);
+        if (network_interfaces_[if_index].ieee1905_socket >= 0) {
+            poller_.Add(network_interfaces_[if_index].ieee1905_socket);
+            poller_.Add(network_interfaces_[if_index].lldp_socket);
```
Here as well, condition must be duplicated.

Yes, it's true that when ieee1905_socket opening failed you also set lldp_socket to -1, but checking them separately is more defensive.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -230,18 +174,23 @@ void Ieee1905Transport::handle_interface_status_change(unsigned int if_index, bo
 
-    if (is_active && network_interfaces_[if_index].fd < 0) {
+    if (is_active && network_interfaces_[if_index].ieee1905_socket < 0) {
         if (!open_interface_socket(if_index)) {
```
Since this function now opens two sockets, perhaps it should be renamed.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -350,7 +299,8 @@ bool Ieee1905Transport::send_packet_to_network_interface(unsigned int if_index,
     packet.header = {.iov_base = &eh, .iov_len = sizeof(eh)};
 
-    int fd             = network_interfaces_[if_index].fd;
+    int fd = packet.ether_type == ETHERTYPE_LLDP ? network_interfaces_[if_index].lldp_socket
```
Maybe this should be a switch and should error out if it is neither.

arnout commented on framework/transport/ieee1905_transport/ieee1905_transport_network.cpp:
```diff
@@ -83,10 +83,14 @@ void Ieee1905Transport::update_network_interfaces(
+    // Note to developer: The current implementation uses AF_PACKET socket with SOCK_RAW protocol
+    // which means we receive
+    // and send packets with the Ethernet header included in the buffer. Please consider changing
+    // implementation to use SOCK_DGRAM protocol (without L2 header handling)
```
I see no reason to switch to SOCK_DGRAM.

fvb-mind committed:
fixup! transport: ieee1905-transport: revert to SOCK_RAW i.s.o. SOCK_DGRAM

bind to bridge interface instead of subinterface
to be seen if this works
